File: libs/east-c/packages/east-c/src/serialization/beast2/sourcemap_table.c
/*
 * BEAST2 v4 Source Map Section.
 *
 * Wire format:
 *   varint(section_byte_length)
 *   varint(stack_count)          -- excludes entry 0 (the empty sentinel)
 *   repeated stack_count times:
 *     varint(frame_count)
 *     repeated frame_count times:
 *       varint(filename_string_idx)
 *       varint(line)
 *       varint(column)
 *
 * Filenames reference the string table for dedup.
 * Entry 0 (empty stack sentinel) is implicit — not written.
 */

#include "internal.h"

/* ================================================================== */
/*  Write source map section                                           */
/* ================================================================== */

void write_source_map_section(Beast2SourceMap *sm, Beast2StringTableEnc *st, ByteBuffer *buf)
{
    ByteBuffer *hdr = byte_buffer_new(256);

    if (!sm || sm->num_stacks <= 1) {
        /* Empty or only sentinel — write stack_count = 0 */
        write_varint(hdr, 0);
    } else {
        /* Skip entry 0 (empty sentinel) */
        write_varint(hdr, sm->num_stacks - 1);
        for (size_t i = 1; i < sm->num_stacks; i++) {
            size_t count = sm->stack_counts[i];
            write_varint(hdr, count);
            for (size_t j = 0; j < count; j++) {
                EastLocation *loc = &sm->stacks[i][j];
                const char *fn = loc->filename ? loc->filename : "";
                size_t fn_len = strlen(fn);
                size_t str_idx = string_table_enc_add(st, fn, fn_len);
                write_varint(hdr, str_idx);
                write_varint(hdr, (uint64_t)loc->line);
                write_varint(hdr, (uint64_t)loc->column);
            }
        }
    }

    write_varint(buf, hdr->len);
    byte_buffer_write_bytes(buf, hdr->data, hdr->len);
    byte_buffer_free(hdr);
}

/* ================================================================== */
/*  Read source map section                                            */
/* ================================================================== */

Beast2SourceMap read_source_map_section(const uint8_t *data, size_t len, size_t *offset,
                                         Beast2StringTableDec *st)
{
    Beast2SourceMap sm = {0};

    uint64_t section_byte_length = read_varint(data, offset);
    size_t section_end = *offset + (size_t)section_byte_length;
    uint64_t stack_count = read_varint(data, offset);

    /* Total entries = stack_count + 1 (for the implicit empty sentinel at 0) */
    sm.num_stacks = (size_t)stack_count + 1;
    sm.stacks = calloc(sm.num_stacks, sizeof(EastLocation *));
    sm.stack_counts = calloc(sm.num_stacks, sizeof(size_t));
    /* Entry 0 is the empty sentinel: stacks[0] = NULL, stack_counts[0] = 0 */

    for (uint64_t i = 0; i < stack_count; i++) {
        uint64_t frame_count = read_varint(data, offset);
        size_t idx = (size_t)i + 1;
        sm.stack_counts[idx] = (size_t)frame_count;
        if (frame_count > 0) {
            sm.stacks[idx] = calloc((size_t)frame_count, sizeof(EastLocation));
            for (uint64_t j = 0; j < frame_count; j++) {
                uint64_t fn_idx = read_varint(data, offset);
                uint64_t line = read_varint(data, offset);
                uint64_t col = read_varint(data, offset);
                if (st && fn_idx < st->count) {
                    sm.stacks[idx][j].filename = strdup(st->strings[fn_idx]);
                }
                sm.stacks[idx][j].line = (int64_t)line;
                sm.stacks[idx][j].column = (int64_t)col;
            }
        }
    }

    *offset = section_end;
    return sm;
}

/* ================================================================== */
/*  Free source map                                                    */
/* ================================================================== */

void beast2_source_map_free(Beast2SourceMap *sm)
{
    if (!sm) return;
    for (size_t i = 0; i < sm->num_stacks; i++) {
        if (sm->stacks[i]) {
            for (size_t j = 0; j < sm->stack_counts[i]; j++) {
                free(sm->stacks[i][j].filename);
            }
            free(sm->stacks[i]);
        }
    }
    free(sm->stacks);
    free(sm->stack_counts);
    sm->stacks = NULL;
    sm->stack_counts = NULL;
    sm->num_stacks = 0;
}

File: libs/east/src/serialization/beast2/v4/sourcemap-table.ts
/**
 * Copyright (c) 2025 Elara AI Pty Ltd
 * Dual-licensed under AGPL-3.0 and commercial license. See LICENSE for details.
 */

import { SourceMap, type Location } from "../../../location.js";
import { BufferWriter, BufferReader } from "../../binary-utils.js";

/**
 * Write the source_map_section in beast2 v3 bespoke format.
 *
 * Wire format:
 *   varint(section_byte_length)
 *   varint(stack_count)
 *   repeated stack_count times:
 *     varint(frame_count)
 *     repeated frame_count times:
 *       varint(filename_string_idx)
 *       zigzag(line)
 *       zigzag(column)
 *
 * If sourceMap is null, writes an empty section (stack_count = 0).
 * Filenames are added to the provided string table for dedup.
 */
export function writeSourceMapSection(
  sourceMap: SourceMap | null,
  stringTable: Map<string, number>,
  writer: BufferWriter,
): void {
  const hw = new BufferWriter();

  if (!sourceMap || sourceMap.size <= 1n) {
    // Empty or only the sentinel empty stack — write stack_count = 0
    hw.writeVarint(0);
  } else {
    const entries = sourceMap.entries();
    // Skip entry 0 (always the empty stack sentinel) — decoder knows index 0 is []
    hw.writeVarint(entries.length - 1);
    for (let i = 1; i < entries.length; i++) {
      const stack = entries[i]!;
      hw.writeVarint(stack.length);
      for (const frame of stack) {
        // Add filename to string table if not present
        let strIdx = stringTable.get(frame.filename);
        if (strIdx === undefined) {
          strIdx = stringTable.size;
          stringTable.set(frame.filename, strIdx);
        }
        hw.writeVarint(strIdx);
        hw.writeVarint(Number(frame.line));
        hw.writeVarint(Number(frame.column));
      }
    }
  }

  const headerBytes = hw.toUint8Array();
  writer.writeVarint(headerBytes.length);
  writer.writeBytes(headerBytes);
}

/**
 * Read the source_map_section from beast2 v3 bespoke format.
 * Returns a SourceMap populated with the decoded stacks.
 */
export function readSourceMapSection(
  reader: BufferReader,
  stringTable: string[],
): SourceMap {
  const sectionByteLength = reader.readVarint();
  const sectionEnd = reader.offset + sectionByteLength;

  const map = new SourceMap();
  const stackCount = reader.readVarint();

  for (let i = 0; i < stackCount; i++) {
    const frameCount = reader.readVarint();
    const stack: Location[] = new Array(frameCount);
    for (let j = 0; j < frameCount; j++) {
      const filenameIdx = reader.readVarint();
      const line = reader.readVarint();
      const column = reader.readVarint();
      stack[j] = {
        filename: stringTable[filenameIdx] ?? '<unknown>',
        line: BigInt(line),
        column: BigInt(column),
      };
    }
    map.intern_stack(stack);
  }

  if (reader.offset !== sectionEnd) {
    throw new Error(`Source map section size mismatch: expected offset ${sectionEnd}, got ${reader.offset}`);
  }

  return map;
}
